db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/BinaryComparisonOperatorNode.java

```diff
@@ -149,31 +149,49 @@ public ValueNode bindExpression(FromList fromList,
 		 */
 		if (! leftTypeId.isStringTypeId() && rightTypeId.isStringTypeId())
 		{
-			DataTypeDescriptor rightTypeServices = rightOperand.getTypeServices();
-
-			rightOperand =  (ValueNode)
-				getNodeFactory().getNode(
-					C_NodeTypes.CAST_NODE,
-					rightOperand, 
-					new DataTypeDescriptor(
-							rightTypeId,
-							true, 
-							rightTypeServices.getMaximumWidth()),
-					getContextManager());
+			TypeId targetTypeId;
+			DataTypeDescriptor targetTypeServices;
+			if (leftTypeId.isBooleanTypeId() || leftTypeId.isDateTimeTimeStampTypeId()) {
+				targetTypeId = leftTypeId;
+				targetTypeServices = leftOperand.getTypeServices();
+			}
+			else {
+				targetTypeId = rightTypeId;
+				targetTypeServices = rightOperand.getTypeServices();
+			}
```

Comment:
String type is only comparable with a boolean or DateTime type, but the error out for other cases is from bindComparisonOperator() which is after here. We can't cast the string to non-boolean/DateTime because it will break the checking in bindComparisonOperator.

I think we should keep the if/else logic. We just need to cast the string type to boolean or DateTime. For other cases (else block), we could raise an error, since it will inevitably raise error out from bindComparisonOperator later.

Reply:
OK, sounds good.